### PR TITLE
Add restart capability to batch driver

### LINKED ISSUE #3765: Restart support in batched derives
**Is your feature request related to a problem? Please describe.**
Batched drivers currently doesn't support dumping restart files. So we need to connect that.
I'm only requesting to output configuration at the end of a driver section.

I think `QMCDriver::finalize` handles the final file I/O.
`QMCDriverNew::finalize` needs to match that. Both random seeds and electron configurations needs to be written to the disk.
HDFWalkerOutput class handles output.

Tests needs to be added in `tests/io/restart`

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

This brings the restart capability in the legacy driver to the batch driver, adding tests `restart_batch-1-16` and `restart_batch-8-2` to verify things are working correctly. Both tests currently fail.

Fixes #3765

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature
- Refactoring (no functional changes, no api changes)
- Testing changes (e.g. new unit/integration/performance tests)
- Documentation changes

### Does this introduce a breaking change?

- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes/No. This PR is up to date with current the current state of 'develop'
- Yes/No. Code added or changed in the PR has been clang-formatted
- Yes/No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes/No. Documentation has been added (if appropriate)

### CODE REVIEW

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
This is work in progress but this is some information that I think its mostly on me for not have documented better in the class declarations.

**Comment by ye-luo** [comment]
Test this please

File: src/QMCDrivers/QMCDriverNew.cpp [hunk 1]
```diff
@@ -392,20 +408,23 @@ void QMCDriverNew::initialLogEvaluation(int crowd_id,
 void QMCDriverNew::setWalkerOffsets()
 {
   std::vector<int> nw(myComm->size(), 0), nwoff(myComm->size() + 1, 0);
-  //  nw[myComm->rank()] = W.getActiveWalkers();
+  population_.saveWalkerConfigurations();
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCDrivers/QMCDriverNew.cpp]: setWalkerOffsets is still needed. I put it static but Its moved to MCPopulation is likely temporary.

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
Test this please

**Review by PDoakORNL — APPROVED** [review]

File: src/QMCDrivers/QMCDriverNew.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCDrivers/QMCDriverNew.cpp]: Still a good incremental step.

**Review by ye-luo — APPROVED** [review]

**Pull request merged**